### ISSUE TITLE
Flushall in Redis Cluster

### ISSUE THREAD

**Issue by skreutzberger** [luin/ioredis#96]
It may just be a lack of understanding but I could not find the answer in browsing the unit tests of a Redis Cluster (latest Redis version).

It seems that a flushall() command which is sent to a cluster connection does take some time to populate or is not properly handled in terms of callback / waiting.

For example, I have the following function which should flush Redis (it is used insde the AfterEach function of my Mocha unit tests) by using Promises (I love Bluebird Promises!):

```js
function flushDBAsync(cluster) {
    log.debug("flushing Redis ...");
    return cluster.flushall()
        .then(function(val) {
            if (val === "OK") {
                log.debug("flushing finished.");
                log.debug("disconnecting from Redis ...");
                return cluster.disconnect();
            } else {
                var msg = "flushing returned with " + val;
                return Promise.reject(msg);
            }
        })
        .then(function() {
            log.debug("disconnecting finished.");
            return true;
        })
        .catch(function(err) {
            log.error("Error in flushDB: " + err);
            return false;
        });
}
```

The function is working always 100% perfectly if I am using a "normal" Redis connection (not a cluster). But when used with a cluster connection then the flushing is not fully completed (populated?) after the return and there are still some keys existing in the cluster.

For now my hacky solution is to wait 2 seconds after I called the function but that is really slowing down my unit tests: `return flushDBAsync(redisConnection).delay(2000);`

How should I properly do a flushall on a Redis Cluster consisting of 3 masters and 3 slaves (default from Redis website)? Redis Cluster is really awesome and I would love to use it with ioredis.

Thanks for looking into that surely very special case!

Sebastian

**Comment by luin** [comment]
Hi Sebastian, ioredis only sends each command to one specified node. If the command has a key, ioredis would calculate the slot of it and send the command to the node serving the slot. However if the command doesn't has a key(e.g. `info`, 'flushall` etc.), it'll be sent to a **random** node of the cluster.

If you want to flush all data of the cluster, you should send `flushall` to all master nodes, which can be accessed via `cluster.masterNodes`(https://github.com/luin/ioredis/blob/master/test/functional/cluster.js#L839):

```javascript
for (var key in cluster.masterNodes) {
  if (!cluster.hasOwnProperty(key)) continue;

  var node = cluster.masterNodes[key];
  node.flushall();
}
```

However it make me confused that you said waiting 2 seconds would fix this, if you can confirm this, I'll definitely look into it.

**Comment by skreutzberger** [comment]
Thanks, your code and explanation made it clearer to me!

I investigated it further using your code and it seems that the connection object just contains masterNodes after 1-2 milliseconds - it is retrieved from Redis right? With your code I could reduce the waiting period from 2 seconds to 2 milliseconds.

So for now the solution is to wait some milliseconds (in my case just 2ms which is fine for unit tests) before flushing a cluster. But that is actually just necessary if you directly flush the cluster after connecting to it (which is unrealistic since some business logic is run in between which should give the connection enough time to get the master nodes).

Here is my new function which flushes all master nodes of a cluster if they are existing:

```js
function flushRedisClusterAsync(cluster) {
    log.debug("flushing Redis Cluster ...");

    var masterNodes = Object.keys(cluster.masterNodes);
    if (masterNodes.length > 0) {
        //log.debug("Found cluster nodes", Object.keys(cluster.nodes));
        log.debug("Found cluster master nodes", masterNodes);

        return Promise.each(masterNodes, function(key) {
            log.debug("flushing cluster master node %s ...", key);
            var node = cluster.masterNodes[key];
            return node.flushall()
            .then(function(val) {
                if(val === "OK") {
                    log.debug("flushed master node", key);
                } else {
                    log.error("Could not flush master node %s. Returned val:", key, val);
                }
            })
        });
    } else {
        log.error("Could not flush cluster, no master nodes were found. 
Please wait some milliseconds after connecting for master node detection!");
        return Promise.reject(false);
    }
}
```

I call that function in my afterEach Mocha function with:

```
return Promise.delay(2)
.then(function(val) {
    return m.flushDBClusterAsync(c);
});
```

I tried it now a lot with a lot of data and it work always 100%. Maybe I could remove the waiting period if I would wait for the "ready" event but actually I don´t know how to add that waiting event to a Mocha beforeEach function (I´m a Node.js newbie coming from Python).

Anyway, I am happy that I can reliably flush a cluster :)

Thanks for your great library, !

**Comment by luin** [comment]
Implemented in 2.0. Closing this issue.

Example:

```javascript
// Send `FLUSHDB` command to all slaves:
var slaves = cluster.nodes('slave');
Promise.all(slaves.map(function (node) {
  return node.flushdb();
}));
```

**Comment by rajatm077** [comment]
Before executing the flushdb command, please check that it is not readOnly node.
```
    const redisNodes = redisClient.nodes();
    for (const node of redisNodes) {
        if (node && node.options && !node.options.readOnly) {
            node.flushdb().catch(err => {
                logger.error('flushRedisKeys-error:', err);
            });
        }
    }
```

**Comment by marnixhoh** [comment]
I noticed that for me it only works if I use `cluster.nodes('master')`. For me all the `slaves` are only in readOnly mode and thus flushing them doesn't work.

I flush Redis after each test suite and this was never really stable. Sometimes it flushed properly and sometimes it didn't. I found the solution to this and thought I'd post it here in case it helps someone else. The trick is to wait for the `ready` event first and then (at least in my case) use the `master` nodes. Below is the code I used:

```
const redisSetup = await new Promise((resolve) => redis.addListener('ready', () => {
    const nodes = redis.nodes('master')
    Promise.all(nodes.map((node) => node.flushdb())).then(resolve)
}))
```

Without waiting for the `ready` event, the nodes returend by `redis.nodes()` is just an empty array.
